### ISSUE TITLE
[template] How to remove a listener registered on createCallback

### ISSUE THREAD

**Issue by koba04** [WICG/webcomponents#708]
TemplateTypeInit has `createCallback` and `processCallback` so we can register an event listener on the `createCallback`.
But how do I able to remove the event listener?
Do we need a new callback like `destroyCallback`, which is called when unmounting the templateInstance?

**Comment by caridy** [comment]
yes, considering that the definition of a template type is global (or at the shadow level), and multiple instances are processed by the same template type processor, we need a way to disconnect these two pieces.

on the example showing `onCheckPoint`, how to disconnect it? What happen when the instance is not longer used?

**Comment by rniwa** [comment]
There is no such a concept as "unmounting". Whatever library / framework that decides to make the template no longer work must manually remove event listeners, etc...

In general, there is not really a concept of deleting a subtree since everything JS can ever reach would always won't be collected by GC. Removing from a document, etc... wouldn't change this.

**Comment by koba04** [comment]
Thanks!

> There is no such a concept as "unmounting". Whatever library / framework that decides to make the template no longer work must manually remove event listeners, etc...

Should we use custom elements or some component systems if we need those lifecycle concepts like mount〜update〜unmount?
(Custom elements have `[connected|disconnected|...]Callback` and that is not a concern of template layer, it's a concern of component layer)

> In general, there is not really a concept of deleting a subtree since everything JS can ever reach would always won't be collected by GC. Removing from a document, etc... wouldn't change this.

Are template instances never released after they are created? I guess defined template type is never released but its instance will be released.
I might be misunderstanding the following point.

> yes, considering that the definition of a template type is global (or at the shadow level), and multiple instances are processed by the same template type processor, we need a way to disconnect these two pieces.

**Comment by rniwa** [comment]
There is no such a thing as *release*. A node is forever accessible as long as some scripts have a reference to it. The same rule applies to template instances.

**Comment by caridy** [comment]
This seems very awkward: 

```js
document.defineTemplateType('self-updating-template', {
    createCallback: function (instance, parts, state) {
        onCheckPoint(() => instance.update(state));
    },
    processCallback: function (instance, parts, state) {
        for (const part of parts)
            part.value = state[part.expression];
    },
});
```

Considering that neither `onCheckPoint`, or `createCallback` or `processCallback` has access to the actual host, nor a way to determine when to release the control. Developers will have to do some gymnastics to add some sort of identity or protocol on the `state` object to release the control when a template instance is not longer useful. e.g.:

```js
document.defineTemplateType('self-updating-template', {
    createCallback: function (instance, parts, state) {
        onCheckPoint(instance, state, (instance, state) => instance.update(state));
    },
    processCallback: function (instance, parts, state) {
        if (state === null) {
             // state set to null is our own protocol which means that the control should be released
             releaseCheckPoint(instance);
             return;
        }
        for (const part of parts)
            part.value = state[part.expression];
    },
});
```

But I understand the constrains here, and the fact that those instances might be created and even updated while still not associated to another element or shadow root. The only thing that comes to my mind to make this easier is to provide another method, equivalent to `update` on the template instance that works on the same fashion by calling the corresponding method on the template type definition so we don't have to hijack the `state` to accomplish that.
